### ISSUE TITLE
rf boosting type for classification

### ISSUE THREAD

**Issue by jmoralez** [microsoft/LightGBM#4118]
## Description

I found that the rf boosting type struggles for classification tasks, even trivial ones. I'd like to know how its `predict_proba` method works, I'd expect it to be the proportion of trees that vote for each class, but I think the implementation may be derived from boosting.

## Reproducible example

```python
import lightgbm as lgb
import matplotlib.pyplot as plt
import numpy as np
from sklearn.datasets import make_blobs

centers = [[-4, -4], [4, 4]]
X, y = make_blobs(n_samples=100, centers=centers, random_state=42)
plt.scatter(X[:, 0], X[:, 1], c=y)
```
This is the data:
![Screenshot from 2021-03-26 22-41-29](https://user-images.githubusercontent.com/8473587/112710122-77b35200-8e84-11eb-8f76-f43b43501644.png)

So a single split at 0 does the job, which is what the trees do:
```python
params = {
    'boosting_type': 'rf',
    'n_estimators': 2,
    'bagging_freq': 1,
    'bagging_fraction': 0.8
}
clf = lgb.LGBMClassifier(**params).fit(X, y)
use_cols = ['tree_index', 'node_depth', 'split_feature', 'threshold', 
            'split_gain', 'decision_type', 'value', 'count']
print(clf.booster_.trees_to_dataframe()[use_cols].to_markdown(index=False))
```
|   tree_index |   node_depth | split_feature   |   threshold |   split_gain | decision_type   |   value |   count |
|-------------:|-------------:|:----------------|------------:|-------------:|:----------------|--------:|--------:|
|            0 |            1 | Column_0        |       1e-35 |       82.988 | <=              |       0 |      83 |
|            0 |            2 |                 |     nan     |      nan     |                 |      -2 |      42 |
|            0 |            2 |                 |     nan     |      nan     |                 |       2 |      41 |
|            1 |            1 | Column_0        |       1e-35 |       79.95  | <=              |       0 |      80 |
|            1 |            2 |                 |     nan     |      nan     |                 |      -2 |      41 |
|            1 |            2 |                 |     nan     |      nan     |                 |       2 |      39 |

However, the probabilities don't reflect that:
```python
prob = clf.predict_proba(X)
prob[:5]
# array([[0.11920292, 0.88079708],
#        [0.88079708, 0.11920292],
#        [0.11920292, 0.88079708],
#        [0.88079708, 0.11920292],
#        [0.11920292, 0.88079708]])
```
So these are basically sigmoid(-2), sigmoid(2).
Adding more trees gives the same result, so I believe the raw scores are being averaged first and then being transformed. Computing the probabilities "manually" does improve when adding more trees:
```python
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

first_tree_scores = clf.predict(X, num_iteration=1, raw_score=True)[:5]
first_tree_scores, sigmoid(first_tree_scores)
# (array([ 2., -2.,  2., -2.,  2.]),
#  array([0.88079708, 0.11920292, 0.88079708, 0.11920292, 0.88079708]))

both_trees_scores = clf.predict(X, raw_score=True)[:5]
both_trees_scores, sigmoid(both_trees_scores)
# (array([ 4., -4.,  4., -4.,  4.]),
#  array([0.98201379, 0.01798621, 0.98201379, 0.01798621, 0.98201379]))
```

## Environment info

LightGBM version or commit hash: 8cc6eefcefb398ce5ae4f0884e12c918c9e4f1e4

**Comment by shiyu1994** [comment]
Hi @jmoralez Yes, the `rf` boosting type uses the same tree learner as `gbdt`. The difference is how the scores are updated. In `gbdt` the prediction values of the new tree are scaled with `learning_rate` and then added to the scores. But in `rf` the scores are simply averaged (divide the sum by the current number of iterations). Note that this averaging operation only happens in training.
https://github.com/microsoft/LightGBM/blob/d517ba12f2e7862ac533908304dddbd770655d2b/src/boosting/rf.hpp#L140-L142
But when predicting, the prediction values of all the trees are simply added up to get the final predicted raw score (without averaging).
https://github.com/microsoft/LightGBM/blob/d517ba12f2e7862ac533908304dddbd770655d2b/src/boosting/gbdt_prediction.cpp#L18-L22

**Comment by jose-moralez** [comment]
Thank you @shiyu1994! The `raw_score` works fine, I'm more concerned about the `predict_proba`. In the example above I'd expect that `sigmoid(clf.predict(X, raw_score=True)) == clf.predict_proba(X)`, however I think the predictions are being averaged in `predict_proba`. https://github.com/microsoft/LightGBM/blob/2885d3e0e4b4a1887316c041d02f3545cb9a2cd0/src/boosting/gbdt_prediction.cpp#L57-L61

So in the example above, no matter how many trees you have, `predict_proba` always returns `sigmoid(-2)` and `sigmoid(2)`, whereas using the `raw_score` you get `num_trees * -2` and `num_trees * 2`.

**Comment by shiyu1994** [comment]
@jose-moralez Thanks for clarifying that. So there's an inconsistence between `predict_proba` and `raw_score`. For this synthetic case, `raw_score` actually gives `100%` accuracy, if we determine the predicted class labels by the signs of the predicted values. As for probability, currently `rf` follows the calculation of `gbdt`. But a better way to calculate the probability for `rf` should be calculating the class proportions in the leaves and then average them.

But currently trees trained in LightGBM don't have the class proportion information in leaves, since we are actually training regression trees. So to implement this, we need to renew the tree leaves after a tree is trained, to calculate the class proportion for `rf`.

I think that requires a non-trivial modification. And perhaps we can take this as a feature request?

**Comment by jmoralez** [comment]
Sorry for the delayed response. Would you suggest applying `sigmoid/softmax` on the raw scores instead of using the `predict_proba` method for this boosting type? I believe with enough trees these can be good enough to pass the tests that we make for classification.

**Comment by shiyu1994** [comment]
@jmoralez Sure, that would be a good work around.

**Comment by jmoralez** [comment]
I believe this can be closed now, since you've solved my doubts about `predict_proba`.
